crates/oxc_semantic/src/scoping.rs

```diff
@@ -99,13 +99,123 @@ impl Default for Scoping {
+/// deallocations involved. Therefore, we store them in an arena allocator to:
+/// 1. Avoid costly heap allocations.
+/// 2. Be able to drop all the data cheaply in one go.
+///
+/// We use [`self_cell`] to be able to store the `Allocator` alongside `Vec`s and `HashMap`s which store
+/// their data in that `Allocator` (self-referential struct).
+///
+/// Conceptually, these structures own their data, and so `ScopingCell` (and therefore also `Scoping`)
+/// should be `Send` and `Sync`, exactly as it would be if `ScopingCell` contained standard heap-allocated
+/// `Vec`s and `HashMap`s.
+///
+/// However, the use of an arena allocator complicates matters, because `Allocator` is not `Sync`,
+/// and `oxc_allocator::Vec` is not `Send`.
+///
+/// ### `Sync`
+///
+/// For it to be safe for `&ScopingCell` to be sent across threads, we must make it impossible to obtain
+/// multiple `&Allocator` references from them on different threads, because those references could be
+/// used to allocate into the same arena simultaneously. `Allocator` is not thread-safe, and this would
+/// likely be undefined behavior.
+///
+/// We prevent this by wrapping the struct created by `self_cell!` in a further wrapper.
+/// That outer wrapper prevents access to methods of `ScopingCellInner` which allow obtaining
+/// an `&Allocator`, unless called with a `&mut self`.
+///
+/// If passed `&mut self`, the method has exclusive access to `ScopingCell`, so no other code can
+/// currently have access to the `Allocator` it contains.
+///
+/// `borrow_allocator` is an unsafe escape hatch to gain access to the `&Allocator` from a `&ScopingCell`,
+/// where the caller must guarantee they do not use that `&Allocator` reference to make any allocations.
+///
+/// ### `Send`
+///
+/// `Allocator` is `Send`. `oxc_allocator::Vec` is not, but that restriction is purely to prevent a `Vec`
+/// being moved to different thread from the `Allocator`, which would allow multiple threads making
+/// allocations in that arena simultaneously.
+///
+/// Here, the `Allocator` and the `Vec`s are contained in the same struct, and moving them to another
+/// thread *together* does not cause a problem.
+///
+/// This is all enclosed in a module, to prevent access to `ScopingCellInner` directly.
+mod scoping_cell {
+    use super::{Allocator, ScopingInner};
+
+    // Inner self-referential struct containing `Allocator` and `ScopingInner`,
+    // where `ScopingInner` contains `Vec`s and `HashMap`s which store their data in the `Allocator`.
+    self_cell::self_cell!(
+        pub struct ScopingCellInner {
+            owner: Allocator,
+            #[covariant]
+            dependent: ScopingInner,
+        }
+    );
+
+    /// Wrapper around [`ScopingCellInner`], which only provides methods that give access to an
+    /// `&Allocator` reference if provided with `&mut ScopingCell`. See comments above.
+    pub struct ScopingCell(ScopingCellInner);
```

Comment:
```suggestion
    #[repr(transparent)]
    pub struct ScopingCell(ScopingCellInner);
```

Reply:
Is there a reason why this *has* to be `#[repr(transparent)]` (e.g. for soundness)? Or are you suggesting this more as general good practice?

Reply:
Just general good practice. Since this is intended to be a new-type wrapper over `ScopingCellInner`, we should request that the compiler does so and doesn't by some miracle of magic decide that, eg. while a `ScopingCellInner` is okay to pass in a register, a `ScopingCell` has to be passed on-stack because "it's a struct mate!".

Reply:
Aapo shared this link which goes into this in more detail: https://www.reddit.com/r/rust/comments/kbyb6z/comment/gflqiep/

Thanks. I was not aware of this at all. There are many other places we should probably be using `#[repr(transparent)]` then e.g.:

https://github.com/oxc-project/oxc/blob/522356201b712f9eb68740362ea2135dbbb7f9c7/crates/oxc_syntax/src/scope.rs#L17

(currently these types are `#[repr(C)]` due to `#[ast]` macro)